### PR TITLE
feat: adopt Alembic for database schema migrations

### PR DESCRIPTION
**Pull request by achilleasatha**
Adopt **Alembic** as the migration framework for `DatabaseSessionService`, replacing manual `ALTER TABLE` operations.

This enables automatic schema tracking, upgrade/downgrade support, and Kubernetes Helm hook integration for production deployments.

Addresses #3343

---

- **AlembicMigrationRunner**
  - Wraps Alembic’s command API with programmatic config (instead of `alembic.ini`)
  - Ships migrations inside the package
- **V1 baseline migration**
  - Creates all current schema tables
- **V0 → V1 in-place bootstrap**
  - Migrates pickle-based databases to JSON
  - No separate destination database required
- `env.py` and `script.py.mako`
  - Support both online and offline migration modes

---

- `upgrade` — Apply pending migrations with auto-bootstrap for existing DBs
- `downgrade` — Roll back by revision spec
- `check` — Exit `0` if up-to-date, `1` if pending (CI/CD gate)
- `stamp` — Bootstrap Alembic tracking for existing databases
- `generate` — Create new migration scripts (autogenerate or template)

---

- `_ensure_alembic_tracking()` stamps baseline after table creation
- V0 databases auto-migrate to V1 on first access
- `ADK_AUTO_MIGRATE_DB=true`
  - Runs pending migrations on startup
- `ADK_AUTO_MIGRATE_DB=false` (default)
  - Raises `RuntimeError` if schema is behind

---

- `get_alembic_revision()`
- `is_alembic_managed()`
- `is_alembic_managed_from_url()`

---

- unit tests for `AlembicMigrationRunner` (SQLite)
- CLI command tests covering all subcommands and edge cases
- integration tests across SQLite / PostgreSQL / MySQL
- Updated `test_database_schema.py` for V0 auto-migration behavior
- GHA workflow:
  - PostgreSQL 15–17
  - MySQL 8.0 / 8.4 / 9.2
  - Python 3.10–3.14
- `docker-compose.yml` for local integration testing / development

---

- `migration_guide.md` — User-facing upgrade and CLI reference
- `helm_migration_guide.md` — Kubernetes Job and Cloud SQL Proxy examples
- `migration/README.md` — Contributor workflow for adding schema versions

**Please ensure you have read the [contribution guide](https://github.com/google/adk-python/blob/main/CONTRIBUTING.md) before creating a pull request.**

### Link to Issue or Description of Change

**1. Link to an existing issue (if applicable):**

- Closes: #_issue_number_
- Related: #_issue_number_

**2. Or, if no issue exists, describe the change:**

_If applicable, please follow the issue templates to provide as much detail as
possible._

**Problem:**
_A clear and concise description of what the problem is._

**Solution:**
_A clear and concise description of what you want to happen and why you choose
this solution._

### Testing Plan

_Please describe the tests that you ran to verify your changes. This is required
for all PRs that are not small documentation or typo fixes._

**Unit Tests:**

- [ ] I have added or updated unit tests for my change.
- [ ] All unit tests pass locally.

_Please include a summary of passed `pytest` results._

**Manual End-to-End (E2E) Tests:**

_Please provide instructions on how to manually test your changes, including any
necessary setup or configuration. Please provide logs or screenshots to help
reviewers better understand the fix._

### Checklist

- [ ] I have read the [CONTRIBUTING.md](https://github.com/google/adk-python/blob/main/CONTRIBUTING.md) document.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have manually tested my changes end-to-end.
- [ ] Any dependent changes have been merged and published in downstream modules.

### Additional context

_Add any other context or screenshots about the feature request here._

### CODE REVIEW

**Comment by achilleasatha** [comment]
Note in this draft version I've included in `/docs` some documentation that should be migrated to https://github.com/google/adk-docs before this is finalised.

Including it here for now so that this is self-contained. 

I am not addressing as part of this PR the locking mechanism discussed in #4387

File: src/google/adk/sessions/alembic_runner.py [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+    branch = row.get("branch")
+    if branch:
+      data["branch"] = branch
+
+    # Unpickle actions
+    actions_raw = row.get("actions")
+    if actions_raw is not None:
+      try:
+        if isinstance(actions_raw, bytes):
+          obj = pickle.loads(actions_raw)  # noqa: S301
```
[the remark this answers is not in the document]
Comment (achilleasatha) [reply src/google/adk/sessions/alembic_runner.py]: Hm not sure what gemini is trying to get out of this. It already acknowledges this is a requirement so we can move away from this. So we will inevitably need to perform the unsafe operation to do the V0->V1 migration. Best I can think of is to explicitly block dangerous types 🤷 

Addressed in commit: 45d77e03a89ad2835abe39951686ce270e758368

File: src/google/adk/sessions/database_session_service.py [hunk 2]
```diff
@@ -236,6 +243,73 @@ async def _prepare_tables(self):
...
+      (no-op if already tracked).
+    """
+    try:
+      from .alembic_runner import AlembicMigrationRunner
+    except ImportError:
+      logger.debug("Alembic not installed; skipping migration tracking.")
+      return
+
+    db_url = str(self.db_engine.url)
+    runner = AlembicMigrationRunner(db_url)
```
Comment (mynewestgitaccount) [inline src/google/adk/sessions/database_session_service.py:349]: It would be better to have the runner use `self.db_engine` directly instead of using the url to create new engines. With that change, kwargs used to create the engine in the db session service init (which can be required for connection) would still be used.

**Comment by ecanlar** [comment]
Hi @wuliang229 👋

Hoping this finds you well! I wanted to follow up on PR #4408 — I see you were assigned as reviewer on Feb 19th and wanted to check if you've had a chance to take a look.

Also, is there a rough timeline for when this might be ready to merge? Happy to address any concerns or make adjustments in the meantime. Thanks!

**Comment by achilleasatha** [comment]
Just giving this a bump as this continues to be a massive pain point for us to date and the teams seems to be adamant about ignoring the need to address this :(

**Comment by ryanovas** [comment]
Can someone from the ADK team at least explain why there is such a resistance to what is such a critical improvement to this framework?
